src/clib/solve_rate_cool_g-cpp.C

```diff
@@ -99,71 +99,60 @@ static void enforce_max_heatcool_subcycle_dt_(
-/// @param[in]     imetal specifies whether or not the caller provided a metal
+/// @param[in]  mask_len the length of the iteration masks
+/// @param[in]  imetal specifies whether or not the caller provided a metal
 ///     density field
-/// @param[in]     min_metallicity specifies the minimum metallicity where we
+/// @param[in]  min_metallicity specifies the minimum metallicity where we
 ///     consider metal chemistry/cooling
-/// @param[in]     ddom specifies precomputed product of mass density and the
+/// @param[in]  ddom specifies precomputed product of mass density and the
 ///    `dom` quantity for each location in `idx_range`
-/// @param[in]     tgas specifies the gas temperatures for the `idx_range`
-/// @param[in]     metallicity specifies the metallicity for the `idx_range`
-///
-/// @todo
-/// It might make more sense to create `itmask_tmp` before calling this
-/// function and then completely ignore the initial values in `itmask` (this
-/// would be far less confusing)
-static void select_chem_scheme_update_masks_(
-  IndexRange idx_range, gr_mask_type* itmask, gr_mask_type* itmask_nr,
-  int* imp_eng, gr_mask_type* itmask_tmp, int mask_len, int imetal,
+/// @param[in]  tgas specifies the gas temperatures for the `idx_range`
+/// @param[in]  metallicity specifies the metallicity for the `idx_range`
+static void setup_chem_scheme_masks_(
+  IndexRange idx_range, const gr_mask_type* itmask, gr_mask_type* itmask_gs,
+  gr_mask_type* itmask_nr, int* imp_eng, int mask_len, int imetal,
   double min_metallicity, const double* ddom, const double* tgas,
   const double* metallicity, const chemistry_data* my_chemistry
 ) {
 
-  std::memcpy(itmask_tmp, itmask, sizeof(gr_mask_type)*mask_len);
+  std::memcpy(itmask_gs, itmask, sizeof(gr_mask_type)*mask_len);
   std::memcpy(itmask_nr, itmask, sizeof(gr_mask_type)*mask_len);
 
   // would it be more robust to use my_chemistry->metal_cooling than imetal?
 
+  // netwon-raphson solver can only coevolves energy when this is true
+  const bool has_nr_coevolve_eint_prereqs = (
+    (my_chemistry->with_radiative_cooling == 1) &&
+    (my_chemistry->primordial_chemistry > 1)
+  );
+
   for (int i = idx_range.i_start; i < idx_range.i_stop; i++) {
-    if ( itmask_tmp[i] != MASK_FALSE )  {
-
-      if ( (imetal == 0) && (ddom[i] < 1.e8) ) {
-        itmask_nr[i] = MASK_FALSE;
-      } else if (
-        (imetal == 1) &&
-        (((metallicity[i] <= min_metallicity) && (ddom[i] < 1.e8)) ||
-         ((metallicity[i] > min_metallicity) && (ddom[i] < 1.e6)))
-      ) {
-        itmask_nr[i] = MASK_FALSE;
-      } else {
-        itmask[i] = MASK_FALSE;
-      }
+    if ( itmask[i] != MASK_FALSE )  {
+      bool usemetal = (imetal == 1) && (metallicity[i] > min_metallicity);
+      bool is_hi_dens = (ddom[i] >= 1.e8) || (usemetal && (ddom[i] >= 1.0e6));
 
-    }
-  }
+      itmask_gs[i] = (!is_hi_dens) ? MASK_TRUE : MASK_FALSE;
+      itmask_nr[i] = (is_hi_dens) ? MASK_TRUE : MASK_FALSE;
```

Comment:
This is a borderline trivial comment, but would it make sense to make it clear that boolean-wise, `itmask_gs` should explicitly have the opposite value to `itmask_nr`? Doing a one-line statement like this that sets one itmask on the other may end up looking a bit clunky, so this might be the simplest after all.